src/local.rs

```diff
@@ -89,25 +106,110 @@ impl From<Error> for super::Error {
     }
 }
 
-/// Local filesystem storage suitable for testing or for opting out of using a
-/// cloud storage provider.
+/// Local filesystem storage providing an [`ObjectStore`] interface to files on
+/// local disk. Can optionally be created with a directory prefix
+///
+/// # Path Semantics
+///
+/// This implementation follows the [file URI] scheme outlined in [RFC 3986]. In
+/// particular paths are delimited by `/`
+///
+/// [file URI]: https://en.wikipedia.org/wiki/File_URI_scheme
+/// [RFC 3986]: https://www.rfc-editor.org/rfc/rfc3986
 #[derive(Debug)]
 pub struct LocalFileSystem {
-    root: std::path::PathBuf,
+    root: Url,
 }
 
 impl std::fmt::Display for LocalFileSystem {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "LocalFileSystem({})", self.root.display())
+        write!(f, "LocalFileSystem({})", self.root)
+    }
+}
+
+impl Default for LocalFileSystem {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl LocalFileSystem {
+    /// Create new filesystem storage with no prefix
+    pub fn new() -> Self {
+        Self {
+            root: Url::parse("file:///").unwrap(),
+        }
+    }
+
+    /// Create new filesystem storage with `prefix` applied to all paths
+    pub fn new_with_prefix(prefix: impl AsRef<std::path::Path>) -> Result<Self> {
+        Ok(Self {
+            root: path_to_url(prefix, true)?,
+        })
+    }
+
+    /// Return filesystem path of the given location
+    fn path_to_filesystem(&self, location: &Path) -> Result<std::path::PathBuf> {
+        // Workaround https://github.com/servo/rust-url/issues/769
+        let mut url = self.root.clone();
+        url.set_path(&format!("{}{}", url.path(), location.to_raw()));
+
+        url.to_file_path()
+            .map_err(|_| Error::InvalidUrl { url }.into())
+    }
+
+    fn filesystem_to_path(&self, location: &std::path::Path) -> Result<Path> {
+        let url = path_to_url(location, false)?;
+        let relative = self.root.make_relative(&url).expect("relative path");
+
+        // TODO: This will double percent-encode
+        Ok(Path::from_raw(relative))
     }
+
+    async fn get_file(&self, location: &Path) -> Result<(fs::File, std::path::PathBuf)> {
+        let path = self.path_to_filesystem(location)?;
+
+        let file = fs::File::open(&path).await.map_err(|e| {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                Error::NotFound {
+                    path: location.to_string(),
+                    source: e,
+                }
+            } else {
+                Error::UnableToOpenFile {
+                    path: path.clone(),
+                    source: e,
+                }
+            }
+        })?;
+        Ok((file, path))
+    }
+}
+
+fn path_to_url(path: impl AsRef<std::path::Path>, is_dir: bool) -> Result<Url, Error> {
```

Comment:
given `path_to_filesystem` and `filesystem_to_path` both are part of `LocalFileSystem`, I suggest making this function also part of `LocalFileSystem` rather than a free function, for consistency (I realize it doesn't actually use `self` technically)

Reply:
I moved it alongside the other free functions